src/database-sqlite3.cpp

```diff
@@ -56,6 +65,50 @@ SQLite format specification:
 			 sqlite3_errmsg(m_database)); \
 	}
 
+static int sqlite3BusyHandler(void *data, int count)
+{
+	s64 &first_time = reinterpret_cast<s64 *>(data)[0];
+	s64 &prev_time = reinterpret_cast<s64 *>(data)[1];
+	s64 cur_time = getTimeMs();
+
+	if (count == 0)
+		first_time = cur_time;
+	else {
+		while (cur_time < prev_time)
+			cur_time += s64(1)<<32;
+	}
+
+	if (cur_time - first_time < BUSY_INFO_TRESHOLD_1) {
+		; // do nothing
+	} else if (cur_time - first_time >= BUSY_INFO_TRESHOLD_1 &&
+			prev_time - first_time < BUSY_INFO_TRESHOLD_1) {
+		infostream << "NOTE: sqlite3 database has been locked for "
+			<< cur_time - first_time << " ms." << std::endl;
+	} else if (cur_time - first_time >= BUSY_INFO_TRESHOLD_2 &&
+			prev_time - first_time < BUSY_INFO_TRESHOLD_2) {
+		infostream << "NOTE: sqlite3 database has been locked for "
+			<< cur_time - first_time << " ms." << std::endl;
+	} else if (cur_time - first_time >= BUSY_ERROR_TRESHOLD &&
+			prev_time - first_time < BUSY_ERROR_TRESHOLD) {
+		errorstream << "WARNING: sqlite3 database has been locked for "
+			<< cur_time - first_time << " ms; this causes lag." << std::endl;
+	} else if (cur_time - first_time >= BUSY_FATAL_TRESHOLD &&
+			prev_time - first_time < BUSY_FATAL_TRESHOLD) {
+		errorstream << "ERROR: sqlite3 database has been locked for "
+			<< cur_time - first_time << " ms - giving up!" << std::endl;
+	} else if ((cur_time - first_time) / BUSY_ERROR_INTERVAL !=
+			(prev_time - first_time) / BUSY_ERROR_INTERVAL) {
+		// Safety net: keep reporting every BUSY_ERROR_INTERVAL
```

Comment:
By 'all this code', do you mean just the last two lines, or the entire if/else if/else if/... ?

In the first case: BUSY_FATAL_TRESHOLD might be changed to be smaller than BUSY_ERROR_INTERVAL (I created another patch to make the times configurable), and it the system freezes for a longer time due to sqlite locks, then a message that keeps appearing (without spamming the user or the logfile) adequately signals the seriousness, and the persistence of the problem.

If you mean the entire is/else if/...: in general, the sqlite lock-wait problem is not impossible, unfortunately. I can easily reproduce it on my system (debian) while mapping. Mapping a non-trivial (larger) world is almost certain to crash minetest at some point in the process. Sometimes the lock delay is larger even than 250ms...

If the lock problem happens, there are just two options: wait or crash. Waiting causes lag, so the players and the administrator need to (or want to) be made aware of it.  Crashing is undesirable, so it should be avoided if possible, at the cost of a longer waiting time, which affects the players...

Returning 0 is a way to make sqlite wrap-up cleanly before returning SQLITE_BUSY to its caller.